src/core/thread/network_data_publisher.cpp

```diff
@@ -216,6 +224,59 @@ void Publisher::NotifyPrefixEntryChange(Event aEvent, const Ip6::Prefix &aPrefix
     mPrefixCallback.InvokeIfSet(static_cast<otNetDataPublisherEvent>(aEvent), &aPrefix);
 }
 
+#if OPENTHREAD_CONFIG_NETDATA_PUBLISHER_OPTIMIZE_ULA_ROUTES
+
+const otIp6Prefix Publisher::kCompactPrefix = {
+    {{{0xfc, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}}},
+    7};
+
+void Publisher::OptimizeUlaRoutes(void)
+{
+    uint8_t         count      = 0;
+    RoutePreference preference = kRoutePreferenceLow;
+
+    for (const PrefixEntry &entry : mPrefixEntries)
+    {
+        if (entry.CanOptimize())
+        {
+            count++;
+            preference = Max(preference, entry.GetPreference());
+        }
+    }
+
+    if (count >= kMinPrefixCountToOptimize)
+    {
+        ExternalRouteConfig config;
+
+        for (PrefixEntry &entry : mPrefixEntries)
+        {
+            if (entry.CanOptimize())
+            {
+                entry.Optimize();
+            }
+        }
+
+        config.Clear();
+        config.mPrefix     = kCompactPrefix;
+        config.mStable     = true;
+        config.mPreference = preference;
+
+        IgnoreError(Get<Local>().AddHasRoutePrefix(config));
+    }
+    else
+    {
+        IgnoreError(Get<Local>().RemoveHasRoutePrefix(AsCoreType(&kCompactPrefix)));
```

Comment:
Thanks @EskoDijk . Yes. The two cases you mention should be covered.
- On second case (`fc00::/7` is itself being published):

  - We remove prefix first, but then it will be re-added from `AddIfOptimized()` call below (the order here is important).
  - We cover it in the test (i.e. we publish `fc00::/7` and then we add/remove other prefixes to go above/below the `N=3` threshold (last steps in `test-019`).
- On first case:
   - It should work with current code (would rely on the fact that publisher would not immediately add prefix). 
   - But to be safer I changed the code to track if/when we add compact prefix (use a new boolean in new push).
   - Also added this situation in the test `test-019`.